File: graphs/python/weighted_graph.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from typing import List, Dict, TypeVar, NamedTuple, Set

# types
VertexType = TypeVar('VertexType',  # Any hashable object
                     str, int, float, complex, str, tuple, frozenset, bytes)
Stack = List


class Edge(NamedTuple):
  """The weighted edge structure."""
  dest: VertexType
  weight: int


class WeightedGraph(object):
  """
  The weighted graph implementation using the adjacency list approach.
  """
  def __init__(self, vertices: List[VertexType]) -> None:
    self._vertices: Dict[VertexType, List[Edge]] = dict((v, []) for v in vertices)
    self._edges: int = 0

  def add_edge(self, src: VertexType, dest: VertexType, weight: int) -> None:
    if src not in self._vertices or dest not in self._vertices:
      return
    self._vertices[src].append(Edge(dest, weight))
    self._edges += 1

  def remove_edge(self, src: VertexType, dest: VertexType) -> None:
    if src not in self._vertices or dest not in self._vertices:
      return
    self._vertices[src] = [e for e in self._vertices[src] if e.dest != dest]
    self._edges -= 1

  def _print_shortest_path(self, path: Dict[VertexType, VertexType], src: VertexType,
                           dest: VertexType) -> None:
    v: VertexType = dest
    res: List[VertexType] = []
    while v != src:
      res.append(v)
      v = path[v]
    res.append(v)
    print(' '.join(str(v) for v in res[::-1]))

  def find_shortest_path(self, src: VertexType, dest: VertexType) -> None:
    """The shortest path finding using the Dijkstra's algorithm."""
    if src not in self._vertices or dest not in self._vertices:
      return
    d: Dict[VertexType, int] = dict.fromkeys(self._vertices.keys(), float('inf'))
    p: Dict[VertexType, VertexType] = {src: src}
    S: Set[VertexType] = {src}
    Q: Set[VertexType] = set(self._vertices.keys())
    d[src] = 0

    while len(Q) > 0:
      # TODO: We should use the fibonacci min-heap here for better performances.
      edge = min([Edge(q, d[q]) for q in Q], key=lambda v: v.weight)
      S.add(edge.dest)
      Q.remove(edge.dest)

      for e in self._vertices[edge.dest]:
        # Relaxation condition.
        if d[e.dest] > d[edge.dest] + e.weight:
          d[e.dest] = d[edge.dest] + e.weight
          p[e.dest] = edge.dest

    self._print_shortest_path(p, src, dest)
    print(f'weight of path: {d[dest]}')

File: graphs/c/weighted_graph.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <limits.h>
#include "weighted_graph.h"

graph_t *create_graph(int vertices)
{
  graph_t *graph = (graph_t *)malloc(sizeof(graph_t));
  if (graph == NULL)
    exit(EXIT_FAILURE);
  graph->edges = 0;
  graph->vertices = vertices;
  graph->lists = (linked_list_t **)malloc(sizeof(linked_list_t *) * graph->vertices);
  if (graph->lists == NULL)
    exit(EXIT_FAILURE);
  for (unsigned int i = 0; i < graph->vertices; i++)
    graph->lists[i] = create_linked_list();
  return graph;
}

void remove_graph(graph_t *g)
{
  assert(g != NULL && g->lists != NULL);
  for (unsigned int i = 0; i < g->vertices; i++)
    remove_linked_list(g->lists[i]);
  free(g);
}

void add_edge(graph_t *g, unsigned int src, unsigned int dest,
              unsigned int weight)
{
  if (src >= g->vertices || dest >= g->vertices)
    return;
  push_back(g->lists[src], dest, weight);
  g->edges++;
}

void remove_edge(graph_t *g, unsigned int src, unsigned int dest)
{
  if (src >= g->vertices || dest >= g->vertices)
    return;
  erase(g->lists[src], dest);
  g->edges--;
}

static inline bool empty_set(bool set[], unsigned int size)
{
  for (unsigned int i = 0; i < size; i++)
    if (set[i] == true)
      return false;
  return true;
}

static inline unsigned int extract_min(bool set[], unsigned int d[],
                                       unsigned int size)
{
  unsigned int min = UINT_MAX;
  for (unsigned int i = 0; i < size; i++)
    if (set[i] == true && d[i] < min)
      min = i;
  return min;
}

static void print_shortest_path(int p[], unsigned int src, unsigned int dest)
{
  while (dest != src) {
    printf("%u ", dest);
    dest = p[dest];
  }
  printf("%u\n", dest);
}

/* The shortest path finding using the Dijkstra's algorithm. */
void find_shortest_path(graph_t *g, unsigned int src, unsigned int dest)
{
  #define UNDEFINED -1
  if (src >= g->vertices || dest >= g->vertices)
    return;
  unsigned int d[g->vertices]; int p[g->vertices];
  bool S[g->vertices], Q[g->vertices];
  for (unsigned int i = 0; i < g->vertices; i++) {
    p[i] = UNDEFINED;  /* Set all parents of vertices to undefined (-1). */
    d[i] = UINT_MAX;   /* Mark distances to vertices as infinity (UINT_MAX). */
    S[i] = false;      /* Clear the set. */
    Q[i] = true;       /* Fill the set with all vertices. */
  }
  d[src] = 0;
  
  while (!empty_set(Q, g->vertices)) {
    /* TODO: We should use the fibonacci min-heap here for better performances. */
    unsigned int min = extract_min(Q, d, g->vertices);
    Q[min] = false; S[min] = true;

    linked_list_node_t *current = g->lists[min]->head;
    for (; current != NULL; current = current->next) {
      /* Relaxation condition. */
      if (d[current->value] > d[min] + current->weight) {
        d[current->value] = d[min] + current->weight;
        p[current->value] = min;
      }
    }
  }

  print_shortest_path(p, src, dest);
  printf("weight of path: %u\n", d[dest]);
  #undef UNDEFINED
}
